[ADP-3327] Revise PKCS7 padding implementation and tests.

jonathanknowles opened the pull request:
## Issue

ADP-3327

## Summary

This PR revises the PKCS#7 padding implementation and property tests.

In particular, we:
- Move functions `pad` and `unpad` into their own dedicated `PKCS7` module.
- Use the `PKCS7` qualifier when importing functions from `PKCS7`.
- Move tests for `pad` and `unpad` into their own dedicated `PKCS7Spec` module.
- Revise the documentation for `pad` and `unpad`.
- Revise the `pad` function to accept zero-length byte strings.
- Add coverage checks to property tests.

jonathanknowles committed:
Move `{pad,unpad}PKCS7` implementation to module `PKCS7`.

Although the PKCS#7 padding scheme is usable with AES 256, it is not
specific to AES 256.

jonathanknowles committed:
Move `{pad,unpad}PKCS7` property tests to module `PKCS7Spec`.

jonathanknowles committed:
Use qualified style for symbols exported from `PKCS7` module.

This removes the need to repeat the term `PKCS7` in the names of
functions that are already within the `PKCS7` namespace.

jonathanknowles committed:
Rename `payload` to `paddedPayload`.

The name is more accurate: the payload is something we must extract
from the padded payload.

jonathanknowles committed:
Actually check that the padding bytes are all identical.

(Instead of assuming that they are all identical.)

jonathanknowles committed:
Fail if the padded payload length is not a multiple of 16 bytes.

jonathanknowles committed:
Fail if the padding is longer than allowed by the specification.

jonathanknowles committed:
Add type signatures to internal definitions of `pad`.

jonathanknowles committed:
Extract out property `prop_pad_unpad`.

We also add coverage checks.

jonathanknowles committed:
Extract out property `prop_pad_length`.

We also add coverage checks.

jonathanknowles committed:
Allow `PKCS7.pad` to accept an empty payload.

This allows us to greatly simplify the properties of `pad` and `unpad`.

The `AES256CBC.encrypt` function is unaffected, and still rejects empty
payloads.

jonathanknowles committed:
Add length-related coverage checks to `prop_pad_length`.

jonathanknowles committed:
Don't give unnecessary prominence to zero-length payloads.

Since the `pad` function is now agnostic to whether or not a payload is
zero-length, we don't need to generate zero-length payloads 50% of the
time.

jonathanknowles committed:
Revise comment for function `PKCS7.pad`.

jonathanknowles committed:
Add comment for function `PKCS7.unpad`.

jonathanknowles committed:
Add property test `prop_pad_isPrefixOf`.

jonathanknowles committed:
Add property `prop_pad_stripPrefix_member_allPossiblePaddingSuffixes`.

jonathanknowles commented on lib/crypto-primitives/src/Cryptography/Cipher/AES256CBC.hs:
```diff
@@ -130,6 +128,9 @@ encrypt mode key iv saltM msg = do
    addSalt salt = saltPrefix <> salt
+   padPKCS7 payload
+       | BS.null payload = Nothing
+       | otherwise = Just (PKCS7.pad payload)
```
This helper function was added to preserve the existing behaviour of `encrypt`, which rejects zero-length payloads.

paweljakubas commented on lib/crypto-primitives/src/Cryptography/Format/PKCS7.hs:
```diff
@@ -0,0 +1,41 @@
+module Cryptography.Format.PKCS7
```
I would be advocate of `Crypthography.Padding.PKCS7` We can have other ways/standards of padding. And Padding better reflects what it is, it is less about presentation, but filling in. Formatting is orthogonal to this concept

jonathanknowles replied:
Good idea -- will change to `Cryptography.Padding.PKCS7`. Thanks for the suggestion!

jonathanknowles replied:
@paweljakubas I've updated this to match your suggestion in 60651241e8766a7a7e92da227a87526b031358a4.

paweljakubas commented on lib/crypto-primitives/src/Cryptography/Format/PKCS7.hs:
```diff
@@ -23,16 +23,16 @@ import qualified Data.ByteString.Char8 as B8
 -- padding. A 16 byte payload requires 2 blocks with padding.
-padPKCS7 :: ByteString -> Maybe ByteString
-padPKCS7 payload
+pad :: ByteString -> Maybe ByteString
```
:+1:

paweljakubas commented on lib/crypto-primitives/crypto-primitives.cabal:
```diff
@@ -54,6 +54,7 @@ library
     , either
     , extra
     , memory                  ^>=0.18
+    , monoid-subclasses
```
:brain:

paweljakubas commented on lib/crypto-primitives/test/Cryptography/Format/PKCS7Spec.hs:
```diff
@@ -26,19 +33,34 @@ import qualified Data.ByteString as BS
+            payload === BS.empty
+        Just paddedPayload ->
+            PKCS7.unpad paddedPayload === Just payload
+    & cover 10
```
:1st_place_medal:

paweljakubas commented on lib/crypto-primitives/test/Cryptography/Format/PKCS7Spec.hs:
```diff
@@ -39,6 +41,19 @@ spec = do
+    & cover 10
+        (BS.length payload `mod` 16 == 0)
+        "BS.length payload `mod` 16 == 0"
+    & cover 10
```
isn't this duplicate of line 63?

paweljakubas commented on lib/crypto-primitives/src/Cryptography/Format/PKCS7.hs:
```diff
@@ -16,16 +16,21 @@ import Data.Semigroup.Cancellative
 -- https://datatracker.ietf.org/doc/html/rfc5652#section-6.3.
--- The padding value is the number of padding bytes.
--- If 1 byte of padding is required, the padding is "01".
--- If 2 bytes of padding, it's "02 02".
```
I like this hands-on examples to be frank ;-)

jonathanknowles replied:
Fair enough! I've added examples in commit 0142b0feeb49aae3c890f37b3ea34a322fc5e088.

What do you think?

paweljakubas commented on lib/crypto-primitives/test/Cryptography/Format/PKCS7Spec.hs:
```diff
@@ -36,6 +44,9 @@ spec = do
     it "prop_pad_length" $
         prop_pad_length
             & property
+    it "prop_pad_stripPrefix_member_allPossiblePaddingSuffixes" $
```
:1st_place_medal:

jonathanknowles committed:
Rename `Cryptography.Format.PKCS7` to `Cryptography.Padding.PKCS7`.

In response to review feedback:

https://github.com/cardano-foundation/cardano-wallet/pull/4528#discussion_r1557488725

paweljakubas reviewed (approved):
very nice PR! I would advocate module renaming, re-checking if we do not have superfluous coverage (16 == 0), also re-consider adding hands-on examples how padding works.

jonathanknowles committed:
Give example padding suffixes.

In response to review feedback:

https://github.com/cardano-foundation/cardano-wallet/pull/4528#pullrequestreview-1988822121
